Simple patches to make jobs robust to database restarts

AlanCoding opened the pull request:
##### SUMMARY
The goal of this is to make it so that a small amount of downtime from postgres (less than 5 seconds) will not fail long-running jobs (multiple hours).

One time in the distant past, this was a design goal, and this goal is obvious if you read the code of `update_model`. Systems have shifted significantly since that time, and several novel failure points have been introduced.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - API


##### ADDITIONAL INFORMATION
My current testing method looks like this:

```bash
echo "select pg_terminate_backend(pid) from pg_stat_activity where datname='awx';" | awx-manage dbshell
```

This is not the _only_ type of database hiccup that might happen, and we should test other things (like a simple restart), this is just my first stop. I have some manual steps, which I will consolidate into an integration test:

 - Launch a job that sleeps for 60 seconds
 - Wait for first lines of output to appear
 - Run the above command to terminate database connections
 - Wait for the rest of the minute to finish
   - **before** this patch the job is marked as failed by the reaper
   - **after** this patch the job completes successfully.

The logs during the event are a little beat up, but the main dispatcher processes remaining running through the whole thing.

AlanCoding committed:
Simple patches to make jobs robust to database restarts

AlanCoding commented on awx/main/dispatch/worker/base.py:
```diff
@@ -152,9 +157,18 @@ def run(self, *args, **kwargs):
+                if not self.pg_is_down:
+                    logger.exception("Postgres error processing tasks")
+                    self.pg_down_time = time.time()
+                    self.pg_is_down = True
```
After this, I believe it would be best to pause for a second or two, so that we're not pummeling a recovering database with failing requests. However, while this is happening we probably need to have a sub-loop checking `self.should_stop` at a faster rate.

AlanCoding replied:
After this, I believe it would be best to pause for a second or two, so that we're not pummeling a recovering database with failing requests. However, while this is happening we probably need to have a sub-loop checking `self.should_stop` at a faster rate.

AlanCoding commented on awx/main/dispatch/worker/base.py:
```diff
@@ -152,9 +157,18 @@ def run(self, *args, **kwargs):
+                    logger.exception("Postgres error processing tasks")
+                    self.pg_down_time = time.time()
+                    self.pg_is_down = True
+                if time.time() - self.pg_down_time > 60.0:
```
We should make the 60 configurable, and maybe we share that constant with `update_model`. But I'm worried about doing any settings references in that cancel_callback (due to threading issues).

AlanCoding replied:
We should make the 60 configurable, and maybe we share that constant with `update_model`. But I'm worried about doing any settings references in that cancel_callback (due to threading issues).

rcdlinux commented:
It would really be great to have this improvement into AWX and Ansible Tower!  A customer is testing some heavy jobs with Tower 3.8 and he has stumbled with this 'issue' with jobs failing after database restarts (or database recovering from a down node to an up node when in a Postgresql cluster).

Congratulations and Thank you very much for providing this patch @AlanCoding !  I hope it can be merged to Awx/Tower soon as possible.

AlanCoding committed:
Add some wait time before retrying loop due to DB error

AlanCoding committed:
Apply dispatcher downtime setting to job updates, fix dispatcher bug

This resolves a bug where the pg_is_down property
  never had the right value
  the loop is normally stuck in the conn.events() iterator
  so it never recognized successful database interactions
  this lead to serial database outages terminating jobs

New setting for allowable PG downtime is shared with task code
  any calls to update_model will use _max_attempts parameter
  to make it align with the patience time that the dispatcher
  respects when consuming new events

AlanCoding commented on awx/main/dispatch/worker/base.py:
```diff
@@ -150,11 +156,27 @@ def run(self, *args, **kwargs):
+                for i in range(10):
+                    if self.should_stop:
+                        return
+                    time.sleep(0.1)
```
This additional logic may be unnecessary because the signal handler ends with `SystemExit`. That may also be making the other checks of `self.should_stop` dead code as well. For defensiveness, I currently opt to keep it.

AlanCoding committed:
To avoid restart loops, handle DB errors on startup with prejudice

brunowcs commented:
Hi @AlanCoding I´m test this code with Patroni HA PostgreSQL with faiolver between node, but only finished my job for me when change time.sleep(0.1) to time.sleep(1).

Job finished, but status is "Task was marked as running but was not present in the job queue, so it has been marked as failed."

I think something here about log_lifecycle.

2022-03-25 20:20:41,349 ERROR    [bc5446138bd44a158a5db67d0962c3a7] awx.main.dispatch Worker failed to run task awx.main.tasks.RunJob(*[15482], **{}
Traceback (most recent call last):
  File "/var/lib/awx/venv/awx/lib64/python3.8/site-packages/awx/main/tasks.py", line 1575, in run
    self.final_run_hook(self.instance, status, private_data_dir, fact_modification_times)
  File "/var/lib/awx/venv/awx/lib64/python3.8/site-packages/awx/main/tasks.py", line 1968, in final_run_hook
    super(RunJob, self).final_run_hook(job, status, private_data_dir, fact_modification_times)
  File "/var/lib/awx/venv/awx/lib64/python3.8/site-packages/awx/main/tasks.py", line 1227, in final_run_hook
    instance.log_lifecycle("finalize_run")
AttributeError: 'NoneType' object has no attribute 'log_lifecycle'

AlanCoding commented on awx/main/dispatch/worker/base.py:
```diff
@@ -150,11 +157,28 @@ def run(self, *args, **kwargs):
+                    self.pg_is_down = True
+                if time.time() - self.pg_down_time > self.pg_max_wait:
+                    logger.warning(f"Postgres event consumer has not recovered in {self.pg_max_wait} s, exiting")
+                    return
```
oh this should be a `raise`, not a `return`

AlanCoding commented:
@brunowcs that error just means to me that it failed to recover in the allotted amount of time. Was database access gone for over 40 seconds? That could explain the behavior you saw.

AlanCoding committed:
If reconnect consistently fails, exit with non-zero code

brunowcs commented:
> @brunowcs that error just means to me that it failed to recover in the allotted amount of time. Was database access gone for over 40 seconds? That could explain the behavior you saw.

Hi Alan, Thanks for this feature. 

my db downtime is very fast (5s until 15s) ,my job print output success, but status not.

2022-03-28 11:47:42,787 ERROR    [ca793dfae4014d94b83d80e368334b6e] awx.main.dispatch Error consuming new events from postgres, will retry for 120 s
Traceback (most recent call last):
  File "/var/lib/awx/venv/awx/lib64/python3.8/site-packages/awx/main/dispatch/worker/base.py", line 158, in run
    for e in conn.events():
  File "/var/lib/awx/venv/awx/lib64/python3.8/site-packages/awx/main/dispatch/__init__.py", line 39, in events
    self.conn.poll()
psycopg2.OperationalError: server closed the connection unexpectedly
        This probably means the server terminated abnormally
        before or while processing the request.

.............
...............
2022-03-28 11:47:45,319 DEBUG    [b552bdc8089c40c2ba98e54d7959e806] awx.analytics.performance request: <WSGIRequest: GET '/'>, response_time: 0.022s
2022-03-28 11:47:47,085 INFO     [-] awx.main.wsbroadcast Active instance with hostname host-001 is registered.
2022-03-28 11:47:47,098 WARNING  [-] awx.main.wsbroadcast Adding {'host-002', 'host-003'} to websocket broadcast list

brunowcs commented:
![tower-error](https://user-images.githubusercontent.com/8129795/160426602-284fc396-f33f-4b45-8e91-1d170be536aa.PNG)

CMPluto commented:
Hi @AlanCoding  & @CFSNM 
On Which AWX release this change has been implemented ?
I am using Awx 21.5.0 but still facing the same issue.

Job finished, but status is "Task was marked as running but was not present in the job queue, so it has been marked as failed."

pls suggest

AlanCoding commented:
^ that message is a catch-all for way too many problems. Honestly we're moving towards architecture changes so this won't happen at all. In these cases, what we firstly seek to know is the receptor status of the job, which is probably killed due to some reason but it is still a struggle to collect that error properly right now.

CMPluto commented:
@AlanCoding 
How can I bring the solution on my existing AWX?
Shall I upgrade the AWX Version from 21.5.0  to latest 21.13.0 version ? 
or the new upcoming AWX version will have the above change in architecture ?
